src/util.cpp

```diff
@@ -337,30 +337,32 @@ void OpenDebugLog()
  * suppress printing of the timestamp when multiple calls are made that don't
  * end in a newline. Initialize it to true, and hold it, in the calling context.
  */
-static std::string LogTimestampStr(const std::string &str, bool *fStartedNewLine)
+static std::string LogTimestampStr(const std::string &str, std::string &logbuf)
 {
-    string strStamped;
-
-    if (!fLogTimestamps)
-        return str;
-
-    if (*fStartedNewLine)
+    if (!logbuf.size())
     {
         int64_t nTimeMicros = GetLogTimeMicros();
-        strStamped = DateTimeStrFormat("%Y-%m-%d %H:%M:%S", nTimeMicros / 1000000);
-        if (fLogTimeMicros)
-            strStamped += strprintf(".%06d", nTimeMicros % 1000000);
-        strStamped += ' ' + str;
+        if (fLogTimestamps)
+        {
+            logbuf = DateTimeStrFormat("%Y-%m-%d %H:%M:%S", nTimeMicros / 1000000);
+            if (fLogTimeMicros)
+                logbuf += strprintf(".%06d", nTimeMicros % 1000000);
+        }
+        logbuf += ' ' + str;
     }
     else
-        strStamped = str;
+    {
+        logbuf += str;
+    }
 
-    if (!str.empty() && str[str.size() - 1] == '\n')
-        *fStartedNewLine = true;
+    if (logbuf.size() && logbuf[logbuf.size() - 1] == '\n')
+    {
+        std::string result = logbuf;
+        logbuf.clear();
+        return result;
```

Comment:
this is a great place to use std::move.  In theory that would move the contents of logbuf into result without making a copy.

This is probably a premature optimization, so this comment is just discussion.